File: multimodal-generation/image-generation/python/06_background_replacement_using_mask_image.py
#!/usr/bin/env python3

from random import randint
from amazon_image_gen import BedrockImageGenerator
import file_utils
import logging
import base64
from datetime import datetime

logging.basicConfig(level=logging.INFO, format='[%(levelname)s] %(message)s')
logger = logging.getLogger(__name__)


def main():
    # Set the image to be edited.
    source_image = "../images/three_pots.jpg"

    # Set the image to be used as a mask. IMPORTANT: The mask must be in PNG format and must contain only pure
    # black and pure white pixels.
    mask_image = "../images/three_pots-center_pot_mask.png"

    # Load the source image from disk.
    with open(source_image, "rb") as image_file:
        source_image_base64 = base64.b64encode(image_file.read()).decode("utf-8")

    with open(mask_image, "rb") as mask_file:
        mask_image_base64 = base64.b64encode(mask_file.read()).decode("utf-8")

    # Configure the inference parameters.
    inference_params = {
        "taskType": "OUTPAINTING",
        "outPaintingParams": {
            "image": source_image_base64,
            "text": "potted flower sitting on a picnic table, backyard wedding decorations, beautiful, highest quality",  # Description of the background to generate
            "maskImage": mask_image_base64,
            "outPaintingMode": "PRECISE",  # "DEFAULT" softens the mask. "PRECISE" keeps it sharp.
        },
        "imageGenerationConfig": {
            "numberOfImages": 1,  # Number of variations to generate. 1 to 5.
            "quality": "standard",  # Allowed values are "standard" and "premium"
            "cfgScale": 7.0,  # How closely the prompt will be followed
            "seed": randint(0, 858993459),  # Use a random seed
        },
    }

    # Define an output directory with a unique name.
    generation_id = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
    output_directory = f"output/{generation_id}"

    # Create the generator.
    generator = BedrockImageGenerator(
        output_directory=output_directory
    )

    # Generate the image(s).
    response = generator.generate_images(inference_params)

    if "images" in response:
        # Save each image
        file_utils.save_base64_images(response["images"], output_directory, "image")

if __name__ == "__main__":
    main()

File: multimodal-generation/image-generation/javascript/06_background_replacement_using_mask_image.js
import { BedrockImageGenerator } from './BedrockImageGenerator.js';
import { imageToBase64 } from './fileUtils.js';

// Path to image to be edited
const sourceImagePath = "../images/three_pots.jpg";

// Set the image to be used as a mask. IMPORTANT: The mask must be in PNG format and must contain only pure
// black and pure white pixels.
const maskImagePath = "../images/three_pots-center_pot_mask.png";

const generateImages = async () => {
    // Format timestamp for unique directory naming
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
    const outputDirectory = `output/${timestamp}`;

    const generator = new BedrockImageGenerator({ outputDirectory });

    try {
        // Read the images from file and encode them as base64 strings
        const sourceImage = await imageToBase64(sourceImagePath);
        const maskImage = await imageToBase64(maskImagePath);

        const params = {
            taskType: 'OUTPAINTING',
            outPaintingParams: {
                image: sourceImage,
                maskImage: maskImage,
                text: "potted flower sitting on a picnic table, backyard wedding decorations, beautiful, highest quality", // Description of the background to generate
                outPaintingMode: "PRECISE",  // "DEFAULT" softens the mask. "PRECISE" keeps it sharp.
            },
            imageGenerationConfig: {
                numberOfImages: 1, // Number of variations to generate. 1 to 5.      
                quality: 'standard', // Allowed values are "standard" and "premium"
                cfgScale: 7.0, // How closely the prompt will be followed
                seed: Math.floor(Math.random() * 858993459), // Use a random seed
            }
        };

        const images = await generator.generateImages(params);
        console.log('Generated images:', images.join(', '));
    } catch (err) {
        console.error('Image generation failed:', err.message);
        process.exit(1);
    }
};

// Self-executing async function
(async () => {
    try {
        await generateImages();
    } catch (err) {
        console.error('Fatal error:', err.message);
        process.exit(1);
    }
})();